### ISSUE TITLE
Self-intersecting Bézier

### ISSUE THREAD

**Issue by hytti** [Pomax/BezierInfo-2#117]
Any cubic bézier curve can be transformed to fit a section of one of six curves:

1. straigt line
2. parabola
3. cubic parabola (one inflection point)
4. semicubical parabola
5. Tschirnhaus(en) cubic (loop)
6. Tschirnhaus(en) cubic secondary form (two inflection points)

I drew a Tschirnhausen family portrait ([tschirn1&2.pdf](https://github.com/Pomax/pomax.github.io/files/1092740/tschirn1.2.pdf)) in which the curves are scaled so that the curvature of the primary Tsirnhausen increases steadily starting from the leftmost point. The two Tschirnhausens and the semicubical converge in infinity.

In the primer (§20: Curve inflections) a nice method for finding the inflection points is described. The quadratic formula without the discriminant gives the location of the "top" point of the bump on the Tschirnhausen secondary. The discriminant tells us the distance (in t) from that point.

The two Tschirnhausen curves are close relatives. Could the intersection on the primary Tschirnhausen be found with a similar method? With the primary form the discriminant is always negative. Let's just multiply it with some negative number to make it positive! The factor -3 seems to give the desired result.

So instead of y²-4xz we have 12xz-3y².

Voilà, .intersects() in Bezier.js becomes straight forward.

_(a misplaced issue - edited and re-posted)_

**Comment by Pomax** [comment]
I'll need to read up on Tschirnhaus forms but that looks pretty interesting!

**Comment by hytti** [comment]
It is a bit difficult to find info on the secondary form but it is mentioned at least here:  
[Newton's diverging parabola](http://www.2dcurves.com/cubic/cubicn.html#tschirnhausen) 

...and this page refers to it by another name (in french):  
[Cubique duplicatrice](https://www.mathcurve.com/courbes2d/cubicduplicatrice/cubicduplicatrice.shtml)

**Comment by Pomax** [comment]
the first link kind of looks like an elliptic curve.

**Comment by hytti** [comment]
my understanding is that they _are_ elliptic curves, only slightly transformed. 4, 5 and 6, that is.

played a bit with grapher:

![tschirnhausen1 2plane](https://user-images.githubusercontent.com/3288498/27625521-9a66b3e6-5bed-11e7-9a5f-13758aaa158f.png)

**Comment by hytti** [comment]
one more thing - there is a nice interactive bézier demo at demos.com. i modified it slightly so that the curve extends beyond 0 < t < 1. playing with the control points is rather hypnotizing and helps you get a grasp of how the loops cusps and bumps behave.

[Bezier Curves extended (original by Christopher Hinton)](https://www.desmos.com/calculator/rd1ltkc5eg)

**Comment by hytti** [comment]
me again! calculated some values for a and b in the elliptic equation. tested them on the [Elliptic Curve Points](https://www.desmos.com/calculator/ialhd71we3) graph i found at desmos.

tschirnhaus i: a=-0.75 b=0.25 _or_ a=-3 b=2 _or_ a=-1 b=0.3849  
tschirnhaus ii: a=-0.75 b=-0.25 _or_ a=-3 b=-2 _or_ a=-1 b=-0.3849  

a^3/b^2 = 27/4

**Comment by Pomax** [comment]
Yeah, the extended interval is useful to look at, I added a section on that to the Primer a while ago, too.

**Comment by hytti** [comment]
that just goes to show how huge a corpus your primer has become! covers every aspect already.
